util/base.hpp

```diff
@@ -0,0 +1,350 @@
+  // \param isBaseFrame is "true" if this is relative to current body frame,
+  // and "false" for using odometry coordinates.
+  static CartesianGoal createFromPosition(Pose p, double goalTime, bool isBaseFrame) {
+    Waypoint start;
+    Waypoint end;
+    start.t = 0.1;
+    start.pos.x = 0;
+    start.pos.y = 0;
+    start.pos.theta = 0;
+
+    end.t = goalTime;
+    end.pos = p;
+
+    createFromWaypoints({start, end}, isBaseFrame);
+  }
+
+  static CartesianGoal createFromWaypoints(std::vector<Waypoint> waypoints, bool isBaseFrame) {
+    auto count = waypoints.size();
+
+    VectorXd times(count);
+    MatrixXd positions(3, count);
+    MatrixXd velocities(3, count);
+    MatrixXd accelerations(3, count);
+
+    for (auto idx = 0; idx < count; ++idx) {
+      times(idx) = waypoints.at(idx).t;
+      positions(0, idx) = waypoints.at(idx).pos.x;
+      positions(1, idx) = waypoints.at(idx).pos.y;
+      positions(2, idx) = waypoints.at(idx).pos.theta;
+
+      velocities(0, idx) = waypoints.at(idx).vel.x;
+      velocities(1, idx) = waypoints.at(idx).vel.y;
+      velocities(2, idx) = waypoints.at(idx).vel.theta;
+
+      accelerations(0, idx) = waypoints.at(idx).acc.x;
+      accelerations(1, idx) = waypoints.at(idx).acc.y;
+      accelerations(2, idx) = waypoints.at(idx).acc.theta;
+    }
+
+    return CartesianGoal(times, positions, velocities, accelerations);
+  }
+
+  // Creates from cartesian trajectory.  May be (x,y) or (x,y,theta), but 
+  // (x,y,theta) less likely to be realizable depending on system.
+  // If "theta" is not defined, is tangent to path.
+  // (x, y path here could be used for planning path through obstacles, for example)
+  // \param isBaseFrame is "true" if this is relative to current body frame,
+  // and "false" for using odometry coordinates.
+  // TODO
+  static CartesianGoal createFromTrajectory(CartesianTrajectory t, bool isBaseFrame);
+
+  const Eigen::VectorXd& times() const { return times_; }
+  const Eigen::Matrix<double, 3, Eigen::Dynamic>& positions() const { return positions_; }
+  const Eigen::Matrix<double, 3, Eigen::Dynamic>& velocities() const { return velocities_; }
+  const Eigen::Matrix<double, 3, Eigen::Dynamic>& accelerations() const { return accelerations_; }
+
+private:
+  CartesianGoal(const Eigen::VectorXd& times,
+       const Eigen::Matrix<double, 3, Eigen::Dynamic>& positions,
+       const Eigen::Matrix<double, 3, Eigen::Dynamic>& velocities,
+       const Eigen::Matrix<double, 3, Eigen::Dynamic>& accelerations)
+    : times_(times),
+      positions_(positions),
+      velocities_(velocities),
+      accelerations_(accelerations) {}
+
+  // TODO: some info here capturing above states/options
+  const Eigen::VectorXd times_{0};
+  const Eigen::Matrix<double, 3, Eigen::Dynamic> positions_{0, 0};
+  const Eigen::Matrix<double, 3, Eigen::Dynamic> velocities_{0, 0};
+  const Eigen::Matrix<double, 3, Eigen::Dynamic> accelerations_{0, 0};
```

Comment:
I wonder about just storing a list of `Waypoint` objects instead, and just converting to the `Eigen` types as needed in the getters?  Probably just moves the complexity rather than reduces it